### DISCUSSION TITLE
AR Model Customization Issue with ar-modes: Default Model Appearing Instead of Custom Design

### DISCUSSION THREAD

**Discussion by abhirami100** [google/model-viewer#4823] [Q&A]
I'm working on a project involving multimodal customization with an option for color customization in AR view. I'm using the ar-modes="scene-viewer quick-look webxr" attribute to support different AR modes.
However, I'm encountering an issue: when I click to view the model in AR, the default model appears instead of the customized design with the selected color. Is there a limitation to using these AR modes, or am I missing something in the implementation? 
![custom](https://github.com/google/model-viewer/assets/110659626/41aadf8f-2ab1-4afc-b0d8-a49f6db31924)

**Comment by danilokrulj** [comment]
If the default model appears that means AR is opened in Scene-Viewer, Scene-Viewer always shows the default model. If you want the changes to be visible in AR just remove scene-viewer and use webxr and quick-look, the changes will be visible in webxr and quick look too since the model will get auto-converted to usdz with the changes u made.

So basically just remove scene-viewer form the ar-modes list and you should be good.

**Reply by danilokrulj** [reply]
Or put webxr in front of scene-viewer so you can fallback to scene-viewer if webxr is not supported for some reason.

**Reply by abhirami100** [reply]
if i remove scene-viewer or put before webxr the custom design is showing.
https://github.com/google/model-viewer/assets/110659626/56b614b1-c7eb-4136-b590-c03823b5b926
When using the scene-viewer here, the default scanning and modeling include true-scale placement. Is there a way to achieve this AR experience while also showcasing custom designs?

**Reply by danilokrulj** [reply]
You wish to display the custom designs in their real scale? I'm just asking if I understood the question correctly.
You can add ar-scale="fixed" to the model-viewer tag to keep the model at 100% scale if that's what you wanted to achieve.

**Reply by abhirami100** [reply]
But when i use ar-scale="fixed" model is showing small in size .But with scene -viewr atribute and give ar-scale="fixed" the model is showing  in actual size.

**Reply by danilokrulj** [reply]
Is the model small when it gets placed or is it small just in the preview (before being placed, when a suitable surface is found the model should be placed in real scale)? You can always check the dimensions in the 3D software you use, it should be its real scale in meters I believe.

**Reply by abhirami100** [reply]
Small when it gets placed.
https://github.com/google/model-viewer/assets/110659626/ecda6416-621a-45db-8100-338686cce150
This video is using a fixed AR scale and does not include a scene viewer.

**Reply by abhirami100** [reply]
@danilokrulj ,  Could you please help me with this question?

**Reply by danilokrulj** [reply]
I'm really not sure what could be the reason. Are you sure the object is placed on the floor and its not a preview? When you start AR you should move your phone all around (left, right, up, down, rotate) until a suitable surface is found, than the object gets placed and it should be 100% scale. 

Maybe you have to update the object scale through code when you change models? But that doesn't sound right to me.

Could you post your code? Maybe that will help.

**Reply by abhirami100** [reply]
import React, { useState, useEffect } from "react";
import "./Multimodel.css";
import car from "../../../assets/car.jpeg";
import table from "../../../assets/table.jpeg";
import chair from "../../../assets/chair.jpeg";
import workout from "../../../assets/workout.jpeg";
import cube from "../../../assets/AR-cube.png";
import TriardLogo from "../../../assets/logo-white.png";
import QRCode from "qrcode.react";

const Multimodel = () => {
	const [selectedModel, setSelectedModel] = useState("Car");
	const [selectedColor, setSelectedColor] = useState(null); // State for selected color
	const [showColorOptions, setShowColorOptions] = useState(true);
	const [qrPopupVisible, setQrPopupVisible] = useState(false);

	const [modelColors] = useState({
		Car: [
			{
				materialName: "Ferrari_550_Barchetta_Pininfarina_2000_by_AlexKa",
				colors: ["#9AC8D6", "#03045e", "#6a040f"],
			},
		],

		desqoomul: [
			{
				sectionName: "Body",
				materialName: "BOARD",
				colors: ["White", "grey", "#966F33"],
			},
			{
				sectionName: "Leg",
				materialName: "LEG",
				colors: ["black", "silver"],
			},
		],
		workout: [
			{
				materialName: "Sliver body",
				colors: ["white", "Grey", "black"],
			},
		],
		Chair: [
			{
				sectionName: "Body",
				materialName: "MAIN BODY",
				colors: ["white", "grey", "#966F33"],
			},
			{
				sectionName: "Leg",
				materialName: "WOOD",
				colors: ["black", "brown"],
			},
		],
	});

	const switchSrc = (name) => {
		setSelectedModel(name);
		setShowColorOptions(true);
	};

	const handleColorSelection = (materialName, color) => {
		setSelectedColor(color); // Update selected color
		applyColorToModel(materialName, color); // Apply color to the model
		console.log(`Selected color for ${selectedModel}: ${color}`);
	};

	const applyColorToModel = (materialName, color) => {
		const modelViewer = document.querySelector("#model-viewer");
		if (modelViewer) {
			const model = modelViewer.model;
			console.log(model);

			if (model) {
				const material = model.materials.find(
					(mat) => mat.name === materialName
				);
				if (material) {
					material.pbrMetallicRoughness.setBaseColorFactor(color);
				}
			}
		}
	};

	const modelImages = {
		Car: car,
		desqoomul: table,
		Chair: chair,
		workout: workout,
	};
	const handleQrPopup = () => {
		setQrPopupVisible(!qrPopupVisible);
	};

	return (
		<div className="multimodel_custom_main">
			<div>
				<div className="header">
					<div className="header_img">
						<a href="http://triard.io/">
							<img src={TriardLogo} alt="Triard Logo" />
						</a>
					</div>
				</div>

				<div className="threed_content2-blender">
					<h3>
						Experience Triard's Innovation:<br></br>Our Pilot Program
					</h3>
					<p style={{ padding: "10px" }}>
						With Triard's 3D configurator, view different models, personalize
						with colors,<br></br> and experience it all in AR.
					</p>
				</div>
			</div>
			<div className="multi_container">
				<div className="model-container">
					<model-viewer
						id="model-viewer"
						src={`/model/${selectedModel}.glb`}
						shadow-intensity="1"
						ar
						camera-controls
						touch-action="pan-y"
						camera-orbit="30deg"
						// ar-modes="scene-viewer quick-look webxr "
						ar-scale="fixed"
						disable-tap
						alt="A 3D model carousel"
						camera-target="-0.000001876m 0.10m -0.03258m"
						min-camera-orbit="auto 32deg auto"
						max-camera-orbit="auto 91deg auto"
					>
						{/* Switch buttons */}
						<div className="slider">
							<div className="slides">
								{Object.keys(modelColors).map((modelName) => (
									<button
										key={modelName}
										className={`slide ${
											selectedModel === modelName && "selected"
										}`}
										onClick={() => switchSrc(modelName)}
										style={{
											backgroundImage: `url(${modelImages[modelName]})`,
											backgroundPosition: "center",
											backgroundSize: "cover",
										}}
									>
										{/* {modelName} */}
									</button>
								))}
							</div>
						</div>
						{/* Color buttons */}
						<div className="multimodel_custom_options">
							{showColorOptions && (
								<div className="color-options">
									{modelColors[selectedModel].map((item, index) => (
										<div key={index}>
											<h3>{item.sectionName}</h3>
											{item.colors.map((color, colorIndex) => (
												<button
													key={colorIndex}
													className="color-button"
													onClick={() =>
														handleColorSelection(item.materialName, color)
													}
													style={{
														backgroundColor: color,
														border: "none",
														cursor: "pointer",
														boxShadow:
															"rgba(50, 50, 93, 0.25) 0px 50px 100px -20px, rgba(0, 0, 0, 0.3) 0px 30px 60px -30px, rgba(10, 37, 64, 0.35) 0px -2px 6px 0px inset",
													}}
												></button>
											))}
										</div>
									))}
								</div>
							)}
							<div
								className="custom_multiar_btn"
								onClick={() => handleQrPopup(0)}
								title="View in AR"
							>
								<img src={cube} alt="AR Cube" /> &nbsp;
							</div>
						</div>
						<button
							slot="ar-button"
							id="ar-button"
							style={{
								backgroundColor: "transparent",
								borderRadius: "4px",
								border: "none",
								position: "absolute",
								right: "224px",
								bottom: "89px",
							}}
						>
							<img src={cube} alt="" />
						</button>

						<div id="ar-prompt">
							<img src="https://invisfurniture.com/_next/image?url=%2F_next%2Fstatic%2Fmedia%2Fthumbcenter.89efb697.png&w=32&q=75" />
						</div>

						<button id="ar-failure">AR is not tracking!</button>
					</model-viewer>
					{qrPopupVisible && (
						<div className="singlepage-qr-popup-multi">
							<div className="qr-content">
								<button
									className="close-button"
									onClick={() => setQrPopupVisible(false)}
								>
									&#10005;
								</button>
								<div className="single-qrpopup">
									<div>
										<QRCode value="https://ar.triard.io/multimodel-customization" />
									</div>
								</div>
								<p>
									Scan QR Code with your smartphone camera to place your model
									in your space.
								</p>
							</div>
						</div>
					)}
				</div>
			</div>
			<div className="custom_multi_contact">
				<h3>Ready to Elevate your E-commerce?</h3>
				<div className="custom_multi_contactbtn">
					<button>
						<a href="https://triard.io/contact-us" target="_blank">
							Contact Us
						</a>
					</button>
				</div>
				<p>
					Join our exclusive pilot program to experience Triard's impact
					firsthand.
				</p>
				<h2>
					Email: <a href="mailto:demo@triard.io">info@triard.io</a>
				</h2>
				<h2>
					Phone: <a href="tel:+91 8594 00 10 10">+91 8594 00 10 10</a>
				</h2>
			</div>
		</div>
	);
};

export default Multimodel;

**Reply by elalish** [reply]
Indeed - do you notice the interaction prompt you put on your page in your video? The little hand moving around in an ellipse? You aren't moving your phone as it is asking you to - when you do, the floor will be found, the model will jump up to true size, and the prompt will disappear. 

Scene Viewer does this a bit faster, but even it won't be accurate until you've moved your phone a bit - the accelerometers are the only thing that tells us how big the space is we're looking at.

**Comment by KPManish** [comment]
Hello i researched a bit about your company , i liked the underlying idea of your startup. Coming back to your query as per my knowledge in scene-viewer you cannot put your custom html, button, design anything. Once the model is launched in ArCore , all it can do is make visible your 3d model without considering textures & texture sliders. While in webxr it is a common issue that model appears smaller or extra bigger, for this you need to go to your 3d software and do trial and error on the sizes until you achieve  what you require.


Hope this helps!!
